Adds automated user creation in production image (#13728)
mistercrunch/Airflow: docs/apache-airflow/production-deployment.rst

Database backend
================

Airflow comes with an ``SQLite`` backend by default. This allows the user to run Airflow without any external
database. However, such a setup is meant to be used for testing purposes only; running the default setup
in production can lead to data loss in multiple scenarios. If you want to run production-grade Airflow,
make sure you :doc:`configure the backend <howto/set-up-database>` to be an external database
such as PostgreSQL or MySQL.

You can change the backend using the following config

.. code-block:: ini

    [core]
    sql_alchemy_conn = my_conn_string

Once you have changed the backend, airflow needs to create all the tables required for operation.
Create an empty DB and give airflow's user the permission to ``CREATE/ALTER`` it.
Once that is done, you can run -

.. code-block:: bash

    airflow db upgrade

``upgrade`` keeps track of migrations already applied, so it's safe to run as often as you need.

.. note::

    Do not use ``airflow db init`` as it can create a lot of default connections, charts, etc. which are not
    required in production DB.

Multi-Node Cluster
==================

> Airflow uses :class:`airflow.executors.sequential_executor.SequentialExecutor` by default. However, by it

Reviewer: ```suggestion
Airflow uses :class:`~airflow.executors.sequential_executor.SequentialExecutor` by default. However, by it
```

Author: Good catch!

nature, the user is limited to executing at most one task at a time. ``Sequential Executor`` also pauses
the scheduler when it runs a task, hence not recommended in a production setup. You should use the

> :class:`Local executor <airflow.executors.local_executor.LocalExecutor>` for a single machine.

Reviewer: ```suggestion
:class:`~airflow.executors.local_executor.LocalExecutor` for a single machine.
```

    --build-arg ADDITIONAL_RUNTIME_APT_DEPS="default-jre-headless"

Actions executed at image start
-------------------------------

If you are using the default entrypoint of the production image,
there are a few actions that are automatically performed when the container starts.
In some cases, you can pass environment variables to the image to trigger some of that behaviour.

Creating system user
....................

Airflow image is Open-Shift compatible, which means that you can start it with random user ID and group id 0.

> Airflow will automatically create such a user and make it's home directoy point to /home/airflow.

Author: ```suggestion
Airflow will automatically create such a user and make it's home directory point to /home/airflow.
```

You can read more about it in `Openshift best practices <https://docs.openshift.com/container-platform/4.1/openshift_images/create-images.html#images-create-guide-openshift_create-images>`_

>  in the "Support arbitrary user ids" chapter.

Author: ```suggestion
in the "Support arbitrary user ids" chapter.
```

Verify Airflow DB connection
............................

In case Postgres or MySQL DB is used, the entrypoint will wait until the airflow DB connection becomes
available. This happens always when you use the default entrypoint.

Upgrading Airflow DB
....................

> If you set ``_AIRFLOW_DB_UPGRADE`` variable to a non-empty value, the entrypoint will perform

Reviewer 2: @potiuk Can you tell me why you decided to prefix the airflow variables with an `_`? Are you just not sure if this is the right way to do this? (i.e are they not worth to be that "public" :D)

If you omit the underscore they're still different to airflow (config) env variables. I just would like to know :)

Author: The main reason is that it is a "behavior" change and it is not image "build" time behavior but container "execution" behaviour. All the other AIRFLOW_* variables impact the way how the image is built, but those have no impact on the image, but change how container behaves.  So I wanted to have them clearly separated so that they are not easily confused. 

Naming is the most difficult part (as usual) - so happy to change it (better idea?) or add some explanation why they are named like that.

WDYT @feluelle ?

Author: Added explanation now, but I am open to any suggestions :)

Reviewer 2: SGTM 👍 Makes sense. Thanks for the explanation.

    --build-arg ADDITIONAL_RUNTIME_APT_DEPS="default-jre-headless"

Actions executed at image start
-------------------------------

If you are using the default entrypoint of the production image,
there are a few actions that are automatically performed when the container starts.
In some cases, you can pass environment variables to the image to trigger some of that behaviour.

The variables that control the "execution" behaviour start with ``_AIRFLOW`` to distinguish them
from the variables used to build the image starting with ``AIRFLOW``.

Creating system user
....................

Airflow image is Open-Shift compatible, which means that you can start it with random user ID and group id 0.

> Airflow will automatically create such a user and make it's home directory point to /home/airflow.

Reviewer 3: ```suggestion
Airflow will automatically create such a user and make it's home directory point to ``/home/airflow``.
```

You can read more about it in `Openshift best practices <https://docs.openshift.com/container-platform/4.1/openshift_images/create-images.html#images-create-guide-openshift_create-images>`_
in the "Support arbitrary user ids" chapter.

Verify Airflow DB connection
............................

In case Postgres or MySQL DB is used, the entrypoint will wait until the airflow DB connection becomes
available. This happens always when you use the default entrypoint.

Upgrading Airflow DB
....................

If you set ``_AIRFLOW_DB_UPGRADE`` variable to a non-empty value, the entrypoint will perform
the ``airflow db upgrade`` command right after verifying the connection. You can also use this

> when you are running airflow with internal sqlite database (default) to upgrade the db and create

Reviewer 3: ```suggestion
when you are running airflow with internal SQLite database (default) to upgrade the db and create
```

+-----------+--------------------------+--------------------------------------------------------------+
| firstname | Airflow                  | _AIRFLOW_WWW_USER_FIRSTNAME                                  |
+-----------+--------------------------+--------------------------------------------------------------+
| lastname  | Admin                    | _AIRFLOW_WWW_USER_LASTNAME                                   |
+-----------+--------------------------+--------------------------------------------------------------+
| email     | airflowadmin@example.com | _AIRFLOW_WWW_USER_EMAIL                                      |
+-----------+--------------------------+--------------------------------------------------------------+
| role      | Admin                    | _AIRFLOW_WWW_USER_ROLE                                       |
+-----------+--------------------------+--------------------------------------------------------------+

In case the password is specified, the user will be attempted to be created, but the entrypoint will
not fail if the attempt fails (this accounts for the case that the user is already created).

> You can, for example start the webserver in the production image with initializing the internal sqlite

Reviewer 3: ```suggestion
You can, for example start the webserver in the production image with initializing the internal SQLite
```

database and creating an ``admin/admin`` Admin user with the following command:

> .. code-block:: bash
> 
>   docker run -it -p 8080:8080 \
>     -e _AIRFLOW_DB_UPGRADE="true" \
>     -e _AIRFLOW_WWW_USER_CREATE="true" \
>     -e _AIRFLOW_WWW_USER_PASSWORD="admin" \
>       apache/airflow:master-python3.8 webserver
> 
> 
> .. code-block:: bash
> 
>   docker run -it -p 8080:8080 \
>     -e _AIRFLOW_DB_UPGRADE="true" \
>     -e _AIRFLOW_WWW_USER_CREATE="true" \
>     -e _AIRFLOW_WWW_USER_PASSWORD_CMD="echo admin" \
>       apache/airflow:master-python3.8 webserver

Reviewer 2: ```suggestion
.. code-block:: bash

  docker run -it -p 8080:8080 \
    -e _AIRFLOW_DB_UPGRADE=true \
    -e _AIRFLOW_WWW_USER_CREATE=true \
    -e _AIRFLOW_WWW_USER_PASSWORD=admin \
      apache/airflow:master-python3.8 webserver

.. code-block:: bash

  docker run -it -p 8080:8080 \
    -e _AIRFLOW_DB_UPGRADE=true \
    -e _AIRFLOW_WWW_USER_CREATE=true \
    -e _AIRFLOW_WWW_USER_PASSWORD_CMD=echo admin \
      apache/airflow:master-python3.8 webserver
```

From the [docs](https://docs.docker.com/engine/reference/commandline/run/#set-environment-variables--e---env---env-file) I would say we remove the `"`. For example I had some issues - not in docker `-e` but in `--env-file` using `"`.

WDYT?

Reviewer 2: https://dev.to/tvanantwerp/don-t-quote-environment-variables-in-docker-268h --> is pretty much the issue I faced, too. But don't know if this also can happen to env variables directly passed into the container.

Author: I think this is only issue specific to .env file, the quotes when you pass environment variables via -e switch are not necessary when you have no spaces (but they are always removed). They are absolutely necessary though when you have spaces (otherwise shell treats what follows as another parameter). This is a "shelll" expansion standard, not  .env file standard and quoting variables there is really good practice.

For example: 

`-e _AIRFLOW_WWW_USER_PASSWORD_CMD=echo admin`

will set `_AIRFLOW_WWW_USER_PASSWORD_CMD` variable to `echo` and it will try to run "admin" image.

Author: This is rather widely used practice.

Author: But indeed probably better to quote all arguments in this case

Author: Updated to add quotes around whole "VAR=value"

> The commands above perform initialization of the sqlite database, create admin user with admin password

Reviewer 3: ```suggestion
The commands above perform initialization of the SQLIte database, create admin user with admin password
```

> and Admin role. They also forward local port 8080 to the webserver port and finally start the webserver.

Reviewer 3: ```suggestion
and Admin role. They also forward local port ``8080`` to the webserver port and finally start the webserver.
```

Verify celery DB connection
...........................

> In case Postgres or MySQL DB is used, and one of the 'scheduler", "celery", "worker", or "flower" commands

Reviewer 3: ```suggestion
In case Postgres or MySQL DB is used, and one of the ``scheduler``, ``celery``, ``worker``, or ``flower`` subcommands
```

there are a few actions that are automatically performed when the container starts.
In some cases, you can pass environment variables to the image to trigger some of that behaviour.

The variables that control the "execution" behaviour start with ``_AIRFLOW`` to distinguish them
from the variables used to build the image starting with ``AIRFLOW``.

Creating system user
....................

Airflow image is Open-Shift compatible, which means that you can start it with random user ID and group id 0.
Airflow will automatically create such a user and make it's home directory point to ``/home/airflow``.
You can read more about it in the "Support arbitrary user ids" chapter in the
`Openshift best practices <https://docs.openshift.com/container-platform/4.1/openshift_images/create-images.html#images-create-guide-openshift_create-images>`_.

Waits for Airflow DB connection
...............................

In case Postgres or MySQL DB is used, the entrypoint will wait until the airflow DB connection becomes
available. This happens always when you use the default entrypoint.

The script detects backend type depending on the URL schema and assigns default port numbers if not specified

> in the URL. Then it loops until connection to the host/port specified can be established

Reviewer: ```suggestion
in the URL. Then it loops until the connection to the host/port specified can be established.
```

It tries ``CONNECTION_CHECK_MAX_COUNT`` times and sleeps ``CONNECTION_CHECK_SLEEP_TIME`` between checks

Supported schemes:

* ``postgres://`` - default port 5432
* ``mysql://``    - default port 3306
* ``sqlite://``

In case of SQLite backend, there is no connection to establish and waiting is skipped.

Upgrading Airflow DB
....................

> If you set ``_AIRFLOW_DB_UPGRADE`` variable to a non-empty value, the entrypoint will perform

Reviewer: ```suggestion
If you set ``_AIRFLOW_DB_UPGRADE`` variable to a non-empty value, the entrypoint will run
```

the ``airflow db upgrade`` command right after verifying the connection. You can also use this
when you are running airflow with internal SQLite database (default) to upgrade the db and create
admin users at entrypoint, so that you can start the webserver immediately. Note - using SQLite is
intended only for testing purpose, never use SQLite in production as it has severe limitations when it
comes to concurrency.

Creating admin user
...................

> The entrypoint can also create www admin user automatically when you enter it. you need to set

Reviewer: ```suggestion
The entrypoint can also create an admin user automatically for the Webserver when you enter it. You need to set
```

Author: I changed it to "webserver user" (to distinguish it from 'system user')

> ``_AIRFLOW_WWW_USER_CREATE`` to a non-empty value in order to do that. This is not intended for
> production, it is only useful if you run a quick test with the production image.

Reviewer: ```suggestion
``_AIRFLOW_WWW_USER_CREATE`` to a non-empty value in order to do that. This is not intended for
production, it is only useful if you would like to run a quick test with the production image.
```

> You need to pass at lest password to create such user via ``_AIRFLOW_WWW_USER_PASSWORD_CMD`` or

Reviewer: ```suggestion
You need to pass at least password to create such a user via ``_AIRFLOW_WWW_USER_PASSWORD_CMD`` or
```

``_AIRFLOW_WWW_USER_PASSWORD_CMD`` similarly like for other ``*_CMD`` variables, the content of
the ``*_CMD`` will be evaluated as shell command and it's output will be set ass password.

User creation will fail if none of the ``PASSWORD`` variables are set - there is no default for
password for security reasons.

+-----------+--------------------------+--------------------------------------------------------------+
| Parameter | Default                  | Environment variable                                         |
+===========+==========================+==============================================================+

> | username  | admin                    | _AIRFLOW_WWW_USER_USERNAME                                   |
> +-----------+--------------------------+--------------------------------------------------------------+
> | password  |                          | _AIRFLOW_WWW_USER_PASSWORD_CMD or _AIRFLOW_WWW_USER_PASSWORD |
> +-----------+--------------------------+--------------------------------------------------------------+
> | firstname | Airflow                  | _AIRFLOW_WWW_USER_FIRSTNAME                                  |
> +-----------+--------------------------+--------------------------------------------------------------+
> | lastname  | Admin                    | _AIRFLOW_WWW_USER_LASTNAME                                   |
> +-----------+--------------------------+--------------------------------------------------------------+
> | email     | airflowadmin@example.com | _AIRFLOW_WWW_USER_EMAIL                                      |
> +-----------+--------------------------+--------------------------------------------------------------+
> | role      | Admin                    | _AIRFLOW_WWW_USER_ROLE                                       |

Reviewer: ```suggestion
| username  | admin                    | ``_AIRFLOW_WWW_USER_USERNAME``                                   |
+-----------+--------------------------+--------------------------------------------------------------+
| password  |                          | ``_AIRFLOW_WWW_USER_PASSWORD_CMD`` or ``_AIRFLOW_WWW_USER_PASSWORD`` |
+-----------+--------------------------+--------------------------------------------------------------+
| firstname | Airflow                  | ``_AIRFLOW_WWW_USER_FIRSTNAME``                                  |
+-----------+--------------------------+--------------------------------------------------------------+
| lastname  | Admin                    | ``_AIRFLOW_WWW_USER_LASTNAME``                                   |
+-----------+--------------------------+--------------------------------------------------------------+
| email     | airflowadmin@example.com | ``_AIRFLOW_WWW_USER_EMAIL``                                      |
+-----------+--------------------------+--------------------------------------------------------------+
| role      | Admin                    | ``_AIRFLOW_WWW_USER_ROLE``                                      |
```

Secured Server and Service Access on Google Cloud
=================================================

This section describes techniques and solutions for securely accessing servers and services when your Airflow
environment is deployed on Google Cloud, or you connect to Google services, or you are connecting
to the Google API.

IAM and Service Accounts
------------------------

> You should do not rely on internal network segmentation or firewalling as our primary security mechanisms.

Reviewer: ```suggestion
You should not rely on internal network segmentation or firewalling as our primary security mechanisms.
```
